src/parser.c

```diff
@@ -2001,84 +2009,141 @@ fgetline (FILE *fp) {
 }
 
-/* Iterate over the log and read line by line (use GNU get_line to parse the
- * whole line).
+/* Parse chunk of lines to logitems */
+void *
+read_lines_thread (void *arg) {
+  GJob *job = (GJob *) arg;
+  for (int i=0; i<job->p; i++) {
+    job->logitems[i] = read_line (job->glog, job->lines[i], &job->test, &job->cnt, job->dry_run);
+
+#ifdef WITH_GETLINE
+    free (job->lines[i]);
+#endif
+  }
+  return (void *) 0;
+}
+
+/* Iterate over the log and read line by line.
+ * With GETLINE: use GNU get_line to parse the whole line.
+ * Without GETLINE: uses a buffer of fixed size.
  *
  * On error, 1 is returned.
  * On success, 0 is returned. */
-#ifdef WITH_GETLINE
 static int
 read_lines (FILE *fp, GLog *glog, int dry_run) {
-  char *line = NULL;
-  int ret = 0, cnt = 0, test = conf.num_tests > 0 ? 1 : 0;
+  int j = 0, ret = 0, cnt = 0, test = conf.num_tests > 0 ? 1 : 0;
+  void *status;
+  GJob jobs[conf.jobs];
+  pthread_t threads[conf.jobs];
+
+#ifndef WITH_GETLINE
+  char *s = NULL;
+#endif
 
   glog->bytes = 0;
-  while ((line = fgetline (fp)) != NULL) {
-    /* handle SIGINT */
-    if (conf.stop_processing)
-      goto out;
-    if ((ret = read_line (glog, line, &test, &cnt, dry_run)))
-      goto out;
-    if (dry_run && NUM_TESTS == cnt)
-      goto out;
-    glog->bytes += strlen (line);
-    free (line);
-    glog->read++;
+
+  for (int k = 0; k < conf.jobs; k++) {
+    jobs[k].p = 0;
+    jobs[k].cnt = 0;
+    jobs[k].glog = glog;
+    jobs[k].test = test;
+    jobs[k].dry_run = dry_run;
   }
 
-  /* if no data was available to read from (probably from a pipe) and
-   * still in test mode, we simply return until data becomes available */
-  if (!line && (errno == EAGAIN || errno == EWOULDBLOCK) && test)
-    return 0;
+  while (!feof (fp)) {
+#ifdef WITH_GETLINE
+    while ((jobs[j].lines[jobs[j].p] = fgetline (fp)) != NULL) {
+#else
+    while ((s = fgets (jobs[j].lines[jobs[j].p], LINE_BUFFER, fp)) != NULL) {
+#endif
+      glog->bytes += strlen (jobs[j].lines[jobs[j].p]);
+      glog->read++;
 
-  return (line && test) || ret || (!line && test && glog->processed);
+      if (++(jobs[j].p) == CHUNK_SIZE)
+        break;
+      j = (j + 1) % conf.jobs;
+    }
 
-out:
-  free (line);
-  /* fails if
-     - we're still reading the log but the test flag was still set
-     - ret flag is not 0, read_line failed
-     - reached the end of file, test flag was still set and we processed lines */
-  return test || ret || (test && glog->processed);
-}
+    for (int k = 0; k < conf.jobs; k++)
+      pthread_create (&threads[k], NULL, read_lines_thread, (void *) &jobs[k]);
+
+    for (int k = 0; k < conf.jobs; k++)
+      pthread_join (threads[k], &status);
+
+#ifdef USE_THIS_VERSION_AFTER_MERGE
+    for (int k = 0; k < conf.jobs; k++) {
+      for (int i = 0; i < jobs[k].p; i++) {
+        if (jobs[k].logitems[i] == NULL)
+          goto out;
+        if (jobs[k].logitems[i]->errstr == NULL)
+          process_log (jobs[k].logitems[i]);
```

Comment:
There appears to be an invalid read (resulting in a segmentation fault) when executing goaccess from the command line:

    ./goaccess access.log -c 
    
It appears to be connected to the fact that, at that moment, it's conducting a dryrun, causing some of the logitems to be null, for instance, `logitem->uniq_key`.

Reply:
Fixed with `if (!dry_run && ...)` 😊